account_credit_control/wizard/credit_control_communication.py

```diff
@@ -0,0 +1,216 @@
+            "       credit_control_policy_level.id)"
+            " WHERE credit_control_line.id in %s"
+            " ORDER by credit_control_policy_level.level, "
+            "          credit_control_line.currency_id"
+        )
+        cr = self.env.cr
+        cr.execute(sql, (tuple(lines.ids),))
+        res = cr.dictfetchall()
+        company_currency = self.env.user.company_id.currency_id
+        datas = []
+        for group in res:
+            data = {}
+            level_lines = self._get_credit_lines(
+                lines.ids,
+                group["partner_id"],
+                group["policy_level_id"],
+                group["currency_id"],
+            )
+            data["credit_control_line_ids"] = [(6, 0, level_lines.ids)]
+            data["partner_id"] = group["partner_id"]
+            data["current_policy_level"] = group["policy_level_id"]
+            data["currency_id"] = group["currency_id"] or company_currency.id
+            datas.append(data)
+        return datas
+
+    @api.model
+    def _generate_comm_from_credit_lines(self, lines):
+        """ Generate a communication object per aggregation of credit lines.
+        """
+        datas = self._aggregate_credit_lines(lines)
+        comms = self.create(datas)
+        return comms
+
+    @api.returns("mail.mail")
+    def _generate_emails(self):
+        """ Generate email message using template related to level """
+        emails = self.env["mail.mail"]
+        required_fields = ["subject", "body_html", "email_from", "email_to"]
+        for comm in self:
+            template = comm.current_policy_level.email_template_id
+            email_values = template.generate_email(comm.id)
+            email_values["message_type"] = "email"
+            # model is Transient record (self) removed periodically so no point
+            # of storing res_id
+            email_values.pop("model", None)
+            email_values.pop("res_id", None)
+            # Remove when mail.template returns correct format attachments
+            attachment_list = email_values.pop("attachments", None)
+            email = emails.create(email_values)
+
+            state = "sent"
+            # The mail will not be send, however it will be in the pool, in an
+            # error state. So we create it, link it with
+            # the credit control line
+            # and put this latter in a `email_error` state we not that we have
+            # a problem with the email
+            if not all(email_values.get(field) for field in required_fields):
+                state = "email_error"
+            comm.credit_control_line_ids.write(
+                {"mail_message_id": email.id, "state": state}
+            )
+            email.attachment_ids = [
+                (
+                    0,
+                    0,
+                    {
+                        "name": att[0],
+                        "datas": att[1],
+                        "datas_fname": att[0],
```

Comment:
Field name is changed to store_fname in v13?

Reply:
Changed, thanks